### ISSUE TITLE
Sorting is not correct when looking for `Profiles: ` commands

### ISSUE THREAD

**Issue by sandy081** [microsoft/vscode#173767]
<img width="687" alt="image" src="https://user-images.githubusercontent.com/10746682/217467117-ec14bec1-836f-4507-a38b-8d582f6e1501.png">

**Comment by bpasero** [comment]
Whats the issue here?

**Comment by bpasero** [comment]
Yes, this is somewhat intentional for commands: we have always sorted matching commands by alphabet to try to group related entries together that share the same prefix and keep the order predictable. We have intentionally not added fuzzy search ranking to commands.

**Comment by TylerLeonhardt** [comment]
People have been hitting this issue more, recently. I wonder why... could be more commands in the command palette making finding the right command challenging

**Comment by isidorn** [comment]
@bpasero this does not look related to fuzzy matching. But shorter titles should get some boost. For example: if my command title has 50 characters and 5 are matching. It should be ranked below a title with 20 characters where 5 are matching.

**Comment by bpasero** [comment]
I cannot find the issue anymore where I explained our reasoning in greater detail, let me post here again: the sorting in the commands picker is historic and we had reasons to do it back then, which I am happy to revisit today if we all agree.

For the longest time, commands picker did not have a list of "most recently used commands" and it was very frustrating to find the right command, given the alphabetic sorting. I added commands history as a way to mitigate that: the history will put results to the top in their order of usage if matching the search query.

![image](https://github.com/microsoft/vscode/assets/900690/17e48663-f51f-4d52-86f8-493854a6b024)

With the idea being that you can build muscle memory very quickly that a query such as "copy" will very likely put the one command to the top that you had always used.

Now, if you leave the history beside and just look at the commands sorting, as you know commands appear in alphabetic order if they match the search input. There is no boost, no fuzzy match, just simple sorting by alphabet. The reason here was to try to keep command results predictable: the order will not change based on your input and similar commands are always appearing together in the same order.

**Comment by isidorn** [comment]
This is getting more important now. Consider this flow: 

https://github.com/user-attachments/assets/d3b18806-3cd6-40f4-86e0-fa0c6a206b0a


Thus, Copilot is sending more "exact matches" to the command palette. And imagine a new user coming to this. We need to make sure that the 100% match is the first result.

**Comment by bpasero** [comment]
If chat knows exactly the command to use, an alternative would be to open command palette with the ID of the command. In that case we only show the matching command.

**Comment by isidorn** [comment]
Well it is AI driven from @vscode, so I am not sure if it can always provide a command ID to use @bhavyaus 
I would still prefer to "fix" the sorting.

**Comment by bhavyaus** [comment]
@bpasero @vscode does have the original command ID, but setting up the quick pick to open with the exact command id results in this:
![Image](https://github.com/user-attachments/assets/f73652a6-5019-43e5-9c3a-965019b55e9a)

So, we use the exact command label instead. Revisiting the sorting would be a better fix here.

**Comment by bpasero** [comment]
I leave it up to Tyler, who owns this. A pragmatic solution would be to force a command to the top when it literally matches the words you typed entirely.

**Comment by TylerLeonhardt** [comment]
@bhavyaus whats the problem with your screenshot? The command you want is at the top, no?

**Comment by hediet** [comment]
In this case, the sorting is so bad, I'd see it as a bug:

![Image](https://github.com/user-attachments/assets/9b548520-4c83-48cf-b54f-928cb04faec0)

If multiple words match, the rank should be the sum of the indexes of all matching words, multiplied by some factor (first index by 1, second index by 2, third by 3 and so on).
Then the results should be sorted ascending by that rank.

**Comment by bhavyaus** [comment]
> [@bhavyaus](https://github.com/bhavyaus) whats the problem with your screenshot? The command you want is at the top, no?

My screenshot was to show why searching via command id when opening the command palette from `@vscode` responses is not the best approach to teaching users about a command. Ideally the sort order would be fixed so that searching via keywords would surface the right commands.

**Comment by hediet** [comment]
I'm running very frequently into this:

![Image](https://github.com/user-attachments/assets/60ea833a-a9e4-4257-9096-e59ba42bbfdb)

I'd suggest to sort them ascending by `2 * item.length + item.indexOf(search), text` when a search is active (or something similar to boost short results and an early occurence of the search). It works very well in the suggest widget, maybe @jrieken has some ideas here?

**Comment by rebornix** [comment]
I understand the original design decision made but with the extensions adding more commands, the command matching experience degrades a ton.

When in a markdown file on github.dev, if you try to find the preview command, this is what you get:

![image](https://github.com/user-attachments/assets/380e7f71-c31f-4293-abcf-009e0f47afa1)

IMHO `command_title.split(word_separators).has(word => word.tolowercase() === query)` should have higher priorities to `p.. re... view` at least.
